src/starter.py

```diff
@@ -0,0 +1,206 @@
+import asyncio
+import os
+import signal
+import sys
+from asyncio.subprocess import Process
+from dataclasses import dataclass
+from pathlib import Path
+from typing import List, Optional
+
+import httpx
+
+from .log import logger
+
+
+@dataclass
+class Service:
+    cmd: List[str]
+    cwd: str
+    timeout: float
+    socket: Optional[str] = None
+    port: Optional[int] = None
+
+
+async def run(
+    cmd: List[str],
+    cwd: Optional[str] = None,
+    pipe_output=False,
+    extra_env: Optional[dict] = None,
+) -> Process:
+    """
+    Run a process
+    :param cmd: Command to run
+    :param cwd: Working directory
+    :param pipe_output: Whether to pipe the output or print it on screen instead
+    :param extra_env: Extra environment variables
+    :return: Process
+    """
+    env = os.environ.copy()
+    if extra_env:
+        env.update(extra_env)
+
+    proc = await asyncio.create_subprocess_exec(
+        *cmd,
+        stdout=asyncio.subprocess.PIPE if pipe_output else sys.stdout,
+        stderr=asyncio.subprocess.PIPE if pipe_output else sys.stderr,
+        cwd=cwd,
+        env=env,
+    )
+
+    return proc
+
+
+async def run_and_wait(
+    cmd: List[str], cwd: Optional[str] = None, extra_env: Optional[dict] = None
+):
+    proc = await run(cmd, cwd, extra_env=extra_env)
+    code = await proc.wait()
+    if code != 0:
+        raise RuntimeError(f"{cmd} failed with code {code}")
+
+
+async def get_output(cmd: List[str], cwd: Optional[str] = None) -> bytes:
+    """
+    Run command and get its stdout if completed successfully
+    :param cmd: Command to run
+    :param cwd: Working directory
+    :return: Bytes from stdout
+    """
+    proc = await run(cmd, cwd, pipe_output=True)
+    out, err = await proc.communicate()
+    if proc.returncode != 0:
+        raise RuntimeError(f"{cmd} failed with code {proc.returncode}")
+    return out
+
+
+async def stop_process(proc: Process, timeout: float = 3.0) -> None:
+    """
+    Try to stop the process gracefully, otherwise killing it
+    :param proc: Process to stop
+    :param timeout: How long to wait for a graceful stop
+    """
+    if proc.returncode is not None:
+        return
+    proc.terminate()
+    await asyncio.wait_for(proc.wait(), timeout=timeout)
+    if proc.returncode is not None:
+        return
+    else:
+        proc.kill()
+
+
+async def setup_nginx() -> None:
+    """
+    Prepare Nginx.
+    Optionally setup HTTP Basic Auth, if environment variables are provided.
+    Process Nginx templates to build actual configuration files.
+    """
+    if user := os.environ.get("HTTP_BASIC_AUTH_USER"):
+        pwd = os.environ["HTTP_BASIC_AUTH_PASSWORD"]
+        realm = os.environ.get("HTTP_BASIC_AUTH_REALM", "Private area")
+        await run_and_wait(
+            ["htpasswd", "-b", "-c", "/run/nginx/.htpasswd", user, pwd],
+            extra_env={"HTTP_BASIC_AUTH_REALM": realm},
```

Comment:
The realm is not needed/used by httpasswd, it's used/filled in by parse-template a bit further down in the call to `get_output`.
> How about instead of setting this to the env here passing it on as an exta env thing to `get_output` and let it pass on to run that does take in extra env stuff.